src/Compilers/CSharp/Portable/Compiler/ClsComplianceChecker.cs

```diff
@@ -158,19 +157,9 @@ public override void VisitNamespace(NamespaceSymbol symbol)
                 CheckMemberDistinctness(symbol);
             }
 
-            if (_compilation.Options.ConcurrentBuild)
+            foreach (var m in symbol.GetMembersUnordered())
```

Comment:
Should we consider using a hybrid approach by changing the original check to:
`if (_compilation.Options.ConcurrentBuild && symbol.IsGlobalNamespace)`?

Reply:
We also need matching changes in VB: http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Compilation/ClsComplianceChecker.vb,97

Reply:
I also noted that we are kicking off parallel foreach for every single named type: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Compiler/ClsComplianceChecker.cs,207. I think that is compounding the allocations.

We ran into similar issues when trying to make the analyzer driver concurrently process compilation events. Attempting to process all events concurrently caused huge allocation and perf hit, and we saw optimal performance with when using a worker queue limited by a certain max worker count: http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerDriver.cs,799. Probably something to consider here?

Reply:
> Should we consider using a hybrid approach

I did investigate an approach where the `VisitNamespace` method was not changed but `VisitNamedType` was. The result showed no improvement over baseline. I stopped looking for hybrid approaches because I couldn't find a scenario that was worse than the one I have here.

Reply:
➡️ Updated the Visual Basic version of the same item.

Reply:
So now we are completely removing the parallelism in this CLS compliance visitor? Seems fine to me as long as there is no build time regression, but probably someone from @dotnet/roslyn-compiler should take a look.

Reply:
Perhaps we should prioritize this: https://github.com/dotnet/roslyn/issues/3679

Reply:
@tmat I left a comment on that issue, suggesting that other scenarios would benefit from such an approach.